wkcuber/converter.py

```diff
@@ -0,0 +1,283 @@
+    if name not in dictionary:
+        dictionary[name] = {value}
+    else:
+        dictionary[name].add(value)
+
+
+def get_source_files(
+    input_path: str, extensions: Iterable[str], allows_single_file_input: bool
+) -> List[str]:
+    source_path = Path(input_path)
+
+    if source_path.is_dir():
+        input_path = path.join(input_path, "**")
+    elif not allows_single_file_input:
+        return []
+
+    source_files = list(find_files(input_path, extensions))
+
+    return source_files
+
+
+class Converter:
+    def accepts_input(self, source_path: str) -> bool:
+        pass
+
+    def convert_input(self, args: Namespace) -> None:
+        pass
+
+
+class WkwConverter(Converter):
+    def accepts_input(self, source_path: str) -> bool:
+        source_files = get_source_files(source_path, {".wkw"}, False)
+        return len(source_files) > 0
+
+    def convert_input(self, args: Namespace) -> None:
+        print("Already a WKW dataset. No conversion necessary...")
+
+
+class NiftiConverter(Converter):
+    def accepts_input(self, source_path: str) -> bool:
+        source_files = get_source_files(source_path, {".nii"}, True)
+        return len(source_files) > 0
+
+    def convert_input(self, args: Namespace) -> None:
+        print("Converting Nifti dataset")
+        convert_nifti(args)
+
+
+class KnossosConverter(Converter):
+    def __init__(self) -> None:
+        self.source_files: List[str] = []
+
+    def accepts_input(self, source_path: str) -> bool:
+        source_files = get_source_files(source_path, {".raw"}, False)
+        self.source_files = source_files
+
+        return len(source_files) > 0
+
+    def convert_input(self, args: Namespace) -> None:
+        print("Converting KNOSSOS dataset")
+
+        # add missing config attributes with defaults
+        put_default_if_not_present(args, "mag", 1)
+        put_default_if_not_present(args, "verbose", True)
+        put_default_if_not_present(args, "jobs", cpu_count())
+        put_default_if_not_present(args, "distribution_strategy", "multiprocessing")
+        put_default_if_not_present(
+            args, "dtype", "uint8"
+        )  # TODO can we autodetect this better?
+
+        (
+            dataset_name,
+            layer_paths_and_mag,
+        ) = self.detect_dataset_and_layer_paths_with_mag()
+
+        for layer_path, mag in layer_paths_and_mag.items():
+            args.source_path = path.join(layer_path, mag)
+            args.layer_name = "color" if layer_path == "" else path.basename(layer_path)
+            args.name = dataset_name
+
+            convert_knossos(args)
+
+    def detect_dataset_and_layer_paths_with_mag(self) -> Tuple[str, Dict[str, str]]:
+        # Path structure for knossos is .../(dataset_name)/(layer_name)/(mag)folder/x0000/y0000/z0000/filename.raw
+        layers: Dict[str, set] = dict()
+        dataset_names = set()
+        for f in self.source_files:
```

Comment:
This is an interesting question, yes it does iterate over all files. Since I already extracted all source files, I didn't even think about other traversal methods.

Reply:
I think, the code would be easier to read when an actual tree traversal is done, but let's discuss this in person.